File: tools/testAngles.py
import math

dimensions = {
    "waist": 110,
    "shoulder": 120,
    "elbow": 160,
    "wrist": 90,
}

def moveTo(x, y, z):
    baseAngle = math.atan2(y, x) * 180 / math.pi

    # Length and height for the shoulder-elbow triangle
    length = math.sqrt(x**2 + y**2)
    height = z - dimensions["waist"] + dimensions["wrist"]
    
    elbowAngleRad = -math.acos((length**2 + height**2 - dimensions["shoulder"]**2 - dimensions["elbow"]**2) / (2 * dimensions["shoulder"] * dimensions["elbow"]))
    elbowAngle = elbowAngleRad * 180 / math.pi

    shoulderAngleRad = math.atan2(height, length) - math.atan2(dimensions["elbow"] * math.sin(elbowAngleRad), dimensions["shoulder"] + dimensions["elbow"] * math.cos(elbowAngleRad))
    shoulderAngle = shoulderAngleRad * 180 / math.pi

    return (baseAngle, elbowAngle, shoulderAngle)

def calculatePosition(baseAngle, elbowAngle, shoulderAngle):
    baseAngleRad = baseAngle * math.pi / 180
    elbowAngleRad = elbowAngle * math.pi / 180
    shoulderAngleRad = shoulderAngle * math.pi / 180

    length = dimensions["shoulder"] * math.cos(shoulderAngleRad) + dimensions["elbow"] * math.cos(shoulderAngleRad + elbowAngleRad)
    height = dimensions["shoulder"] * math.sin(shoulderAngleRad) + dimensions["elbow"] * math.sin(shoulderAngleRad + elbowAngleRad)

    x = length * math.cos(baseAngleRad)
    y = length * math.sin(baseAngleRad)
    z = height + dimensions["waist"] - dimensions["wrist"]

    return (x, y, z)

angles = moveTo(108, 15, 50)
print(f"Base: {angles[0]}")
print(f"Shoulder: {angles[2]}")
print(f"Elbow: {angles[1]}")

position = calculatePosition(angles[0], angles[1], angles[2])
print(f"X: {position[0]}")
print(f"Y: {position[1]}")
print(f"Z: {position[2]}")

File: tools/testAngles.cpp
#include <iostream>
#include <cmath>

#define PI 3.14159265
#define armDimensions { 110, 115, 155, 90 }

typedef struct {
    float waist_length;
    float shoulder_length;
    float elbow_length;
    float wrist_length;
} RoboticArmDimensions;

RoboticArmDimensions dimensions = armDimensions;

void moveTo(float x, float y, float z) {
    // Base angle is calculated by the angle between the origin and the point (x, y)
    float baseAngle = atan2(y, x) * 180 / PI;

    // Horizontal length required for the elbow + shoulder to reach the desired point
    float length = sqrt(x * x + y * y);
    // Vertical height required for the elbow + shoulder to reach the desired point
    float height = z - dimensions.waist_length + dimensions.wrist_length;

    // Law of cosines to calculate the angle between the shoulder and the elbow

    float elbowAngleRad = -acos((length * length + height * height - dimensions.shoulder_length * dimensions.shoulder_length - dimensions.elbow_length * dimensions.elbow_length ) / (2 * dimensions.shoulder_length * dimensions.elbow_length));
    float shoulderAngleRad = atan2(height, length) - atan2(dimensions.elbow_length * sin(elbowAngleRad), dimensions.shoulder_length + dimensions.elbow_length * cos(elbowAngleRad));

    // Convert radians to degrees
    float elbowAngle = elbowAngleRad * 180 / PI;
    float shoulderAngle = shoulderAngleRad * 180 / PI;

    // Wrist should be perpendicular to the ground
    float wristAngle = -90 - shoulderAngle - elbowAngle;

    std::cout << "Base angle: " << baseAngle << std::endl;
    std::cout << "Shoulder angle: " << shoulderAngle << std::endl;
    std::cout << "Elbow angle: " << elbowAngle << std::endl;
    std::cout << "Wrist angle: " << wristAngle << std::endl;
}

int main() {
    float x = 50;
    float y = 200;
    float z = 100;

    moveTo(x, y, z);
}